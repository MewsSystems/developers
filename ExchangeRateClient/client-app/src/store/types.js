// @flow strict

export type Data = {|
  [id: string]: {|
    currencies: Array<{| +code: string, +name: string |}>,
  |},
|};

export type SelectedRates = string[];

export type Rates = {| [id: string]: {| +current: number, +before: number |} |};

export type StateTypes = {
  +isFetchingConfig: boolean,
  +isFetchingRates: boolean,
  +rates: Rates,
  +selectedRates: SelectedRates,
  +data: Data,
  +APIerror: ?Error,
};
