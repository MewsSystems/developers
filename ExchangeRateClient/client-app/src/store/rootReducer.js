// @flow

import {
  CONFIG_FETCH_SUCCESS,
  CONFIGS_FETCH_START,
  CONFIGS_FETCH_FAIL,
  RATES_FETCH_START,
  RATES_FETCH_SUCCESS,
  RATES_FETCH_FAIL,
  SELECT_RATES_IDS,
  type Action,
} from "../actions/configActions";
import type { StateTypes } from "./types";

export const initalState: StateTypes = {
  selectedRates: [],
  data: {},
  isFetchingConfig: false,
  APIerror: null,
  APIerror: null,
  isFetchingRates: false,
  rates: {},
};

export default (state: StateTypes = { ...initalState }, action: Action) => {
  switch (action.type) {
    case CONFIG_FETCH_SUCCESS:
      return {
        ...state,
        isFetchingConfig: false,
        ...action.payload,
      };
    case CONFIGS_FETCH_START:
      return {
        ...state,
        isFetchingConfig: true,
      };
    case CONFIGS_FETCH_FAIL:
      return {
        ...initalState,
        APIerror: action.payload.error,
      };
    case RATES_FETCH_START:
      return {
        ...state,
        selectedRates: action.payload,
        isFetchingRates: true,
      };
    case RATES_FETCH_SUCCESS: {
      const rates = action.payload;
      const ids = Object.keys(rates);

      const temp = ids.reduce(
        (acc, id) => ({
          ...acc,
          [id]: {
            current: rates[id],
            before: state.rates && state.rates[id] && state.rates[id].current,
          },
        }),
        {},
      );

      return {
        ...state,
        rates: temp,
        isFetchingRates: false,
      };
    }
    case RATES_FETCH_FAIL:
      return {
        ...state,
        isFetchingRates: false,
        APIerror: action.payload.error,
      };

    case SELECT_RATES_IDS:
      return {
        ...state,
        selectedRates: action.payload.ids,
      };

    default:
      return state;
  }
};
