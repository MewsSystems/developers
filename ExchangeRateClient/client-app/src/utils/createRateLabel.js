// @flow

export default (currencies: Array<{ code: string, name: string }>) =>
  `${currencies[0].code} - ${currencies[1].code}`;
