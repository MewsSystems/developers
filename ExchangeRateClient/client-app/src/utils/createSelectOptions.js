// @flow strict

// TODO name ad to const
const DEFAULT_SELECTED_COUNT = 3;

export const sanitizeConfiguration = ({
  currencyPairs,
}: {
  currencyPairs: {
    [id: string]: {
      code: string,
      name: string,
    },
  },
}) => {
  const ids: string[] = Object.keys(currencyPairs);

  const temp = ids.reduce(
    (acc, id, index) => ({
      ...acc,
      [id]: { currencies: currencyPairs[id], isSelected: index < DEFAULT_SELECTED_COUNT && true },
    }),
    {},
  );

  const result = {
    selectedRates: ids,
    data: temp,
  };
  return result;
};
