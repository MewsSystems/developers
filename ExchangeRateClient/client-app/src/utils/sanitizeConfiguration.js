// @flow strict

export const sanitizeConfiguration = ({
  currencyPairs,
}: {
  currencyPairs: {
    [id: string]: {
      code: string,
      name: string,
    },
  },
}) => {
  const ids: string[] = Object.keys(currencyPairs);

  const temp = ids.reduce(
    (acc, id, index) => ({
      ...acc,
      [id]: { currencies: currencyPairs[id] },
    }),
    {},
  );

  const result = {
    selectedRates: ids,
    data: temp,
  };
  return result;
};
