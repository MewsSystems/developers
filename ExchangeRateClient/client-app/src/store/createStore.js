// @flow strict

import { createStore, applyMiddleware } from "redux";
import thunk from "redux-thunk";
import { composeWithDevTools } from "redux-devtools-extension";
import rootReducer from "./rootReducer";

export default function configureStore() {
  return createStore(rootReducer, composeWithDevTools(applyMiddleware(thunk)));
}
