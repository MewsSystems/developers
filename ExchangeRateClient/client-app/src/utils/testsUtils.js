// @flow
import * as React from "react";

import { Provider, connect } from "react-redux";
import { render } from "react-testing-library";
import { createStore } from "redux";
import rootReducer from "../store/rootReducer";

export function renderWithRedux(
  ui: React.Node,
  {
    initialState,
    store = createStore(rootReducer, initialState),
  }: { initialState?: Object, store?: Object } = {},
) {
  return {
    ...render(<Provider store={store}>{ui}</Provider>),
    store,
  };
}
