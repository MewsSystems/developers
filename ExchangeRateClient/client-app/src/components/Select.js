// @flow strict

import * as React from "react";
import Select from "react-select";
import { connect } from "react-redux";
import config from "../records/config";
import { fetchRatesAction, selectIds } from "../actions/index";
import type { SelectedRates, Data } from "../store/types";
import type { ThunkAction } from "../actions/index";

import createRateLabel from "../utils/createRateLabel";

const customStyles = {
  option: (base, state) => ({
    ...base,
    backgroundColor: state.isFocused ? "rgba(249, 222, 201, 1)" : base.backgroundColor,
    color: "rgba(58, 64, 90, 1)",
  }),
  control: (base, state) => ({
    ...base,
    backgroundColor: "transparent",
    boxShadow: "inset -11px -15px 84px -32px rgba(0,0,0,1)",
  }),
  multiValue: (base, state) => {
    const opacity = state.isDisabled ? 0.5 : 1;
    const transition = "opacity 300ms";
    return { ...base, opacity, transition, fontSize: "1.5em" };
  },
};

type Props = {
  data: Data,
  selectedRates: SelectedRates,
  fetchRatesAction: (ids: string[]) => ThunkAction,
  selectIds: (ids: string[]) => { type: "SELECT_RATES_IDS", payload: { ids: string[] } },
};

const CustomSelect = (props: Props) => {
  if (Object.keys(props.data).length === 0) return null;
  const { data, selectedRates, fetchRatesAction, selectIds } = props;

  const opt = Object.keys(data).reduce((acc, id) => {
    const item = {
      value: id,
      label: createRateLabel(data[id].currencies),
    };

    return [...acc, item];
  }, []);

  return (
    <div data-testid="select">
      <Select
        styles={customStyles}
        isMulti
        onChange={data => selectIds(data.map(({ value }) => value))}
        options={opt}
        defaultValue={opt.filter(item => selectedRates.includes(item.value))}
      />
    </div>
  );
};

const mapStateToProps = ({ data = {}, selectedRates = [] }) => ({
  data,
  selectedRates,
});

export default connect(
  mapStateToProps,
  { fetchRatesAction, selectIds },
)(CustomSelect);
