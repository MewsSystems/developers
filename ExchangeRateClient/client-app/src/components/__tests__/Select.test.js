// @flow
import { render, cleanup } from "react-testing-library";
import React from "react";
import Select from "react-select";
import MySelect from "../Select";
import { renderWithRedux } from "../../utils/testsUtils";
import store from "../../records/storeState";

afterEach(cleanup);

jest.mock("react-select", () => props => {
  expect(props).toMatchSnapshot();
  return null;
});

describe("Select", () => {
  test("test if props are set properly", () => {
    const { baseElement } = renderWithRedux(<MySelect />, { initialState: store });
    expect(baseElement).toMatchSnapshot();
  });
});
