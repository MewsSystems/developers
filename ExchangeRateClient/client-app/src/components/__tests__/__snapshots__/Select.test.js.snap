// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select test if props are set properly 1`] = `
Object {
  "defaultValue": Array [
    Object {
      "label": "AMD - GEL",
      "value": "70c6744c-cba2-5f4c-8a06-0dac0c4e43a1",
    },
    Object {
      "label": "PHP - DZD",
      "value": "41cae0fd-b74d-5304-a45c-ba000471eabd",
    },
    Object {
      "label": "LSL - AUD",
      "value": "5b428ac9-ec57-513d-8a08-20199469fb4d",
    },
    Object {
      "label": "BDT - VEF",
      "value": "f816e384-0e43-5ce7-a017-deaa8d666774",
    },
    Object {
      "label": "QAR - JOD",
      "value": "5b98842f-bfe5-5564-b321-068763d7e2a3",
    },
    Object {
      "label": "SRD - CZK",
      "value": "a2bda952-4369-5d41-9d0b-e6c9947e5b82",
    },
    Object {
      "label": "CZK - LSL",
      "value": "1993f7b9-f9be-551a-beac-312d6befd0cf",
    },
    Object {
      "label": "DZD - MVR",
      "value": "61fb0e0d-626e-5e0a-831a-ef95d5c32203",
    },
    Object {
      "label": "BIF - CLP",
      "value": "b7fdd67f-5051-58b7-a3c6-84f5da637df5",
    },
    Object {
      "label": "HKD - FKP",
      "value": "611398c5-6bd9-596e-8803-3ed0b093995d",
    },
  ],
  "isMulti": true,
  "onChange": [Function],
  "options": Array [
    Object {
      "label": "AMD - GEL",
      "value": "70c6744c-cba2-5f4c-8a06-0dac0c4e43a1",
    },
    Object {
      "label": "PHP - DZD",
      "value": "41cae0fd-b74d-5304-a45c-ba000471eabd",
    },
    Object {
      "label": "LSL - AUD",
      "value": "5b428ac9-ec57-513d-8a08-20199469fb4d",
    },
    Object {
      "label": "BDT - VEF",
      "value": "f816e384-0e43-5ce7-a017-deaa8d666774",
    },
    Object {
      "label": "QAR - JOD",
      "value": "5b98842f-bfe5-5564-b321-068763d7e2a3",
    },
    Object {
      "label": "SRD - CZK",
      "value": "a2bda952-4369-5d41-9d0b-e6c9947e5b82",
    },
    Object {
      "label": "CZK - LSL",
      "value": "1993f7b9-f9be-551a-beac-312d6befd0cf",
    },
    Object {
      "label": "DZD - MVR",
      "value": "61fb0e0d-626e-5e0a-831a-ef95d5c32203",
    },
    Object {
      "label": "BIF - CLP",
      "value": "b7fdd67f-5051-58b7-a3c6-84f5da637df5",
    },
    Object {
      "label": "HKD - FKP",
      "value": "611398c5-6bd9-596e-8803-3ed0b093995d",
    },
  ],
  "styles": Object {
    "control": [Function],
    "multiValue": [Function],
    "option": [Function],
  },
}
`;

exports[`Select test if props are set properly 2`] = `
<body>
  <div />
</body>
`;
